beacon-chain/blockchain/process_block.go

```diff
@@ -518,11 +518,33 @@ func (s *Service) isDataAvailable(ctx context.Context, root [32]byte, signed int
 	if err != nil {
 		return errors.Wrap(err, "could not get KZG commitments")
 	}
-	sidecars, err := s.cfg.BeaconDB.BlobSidecarsByRoot(ctx, root)
-	if err != nil {
-		return errors.Wrap(err, "could not get blob sidecars")
+	// wait until we have received all blobs for this block
+	s.blobNotifier.Lock()
+	var nc blobNotifierChan
+	var ok bool
+	nc, ok = s.blobNotifier.chanForRoot[root]
+	if !ok {
+		nc = blobNotifierChan{indices: []uint64{}, channel: make(chan struct{})}
+		s.blobNotifier.chanForRoot[root] = nc
+	}
+	s.blobNotifier.Unlock()
+	for {
+		select {
+		case <-nc.channel:
+			if len(kzgCommitments) == len(nc.indices) {
```

Comment:
Did you mean to make chanForRoot a `map[[32]*blobNotifierChan`? As currently written, `nc.indices` will never be updated by the blob side, because the read from the map makes a copy of the `blobNotifierChan` struct, so the slice append modifies the local copy. Using references on both sides will fix that.

Reply:
for thread safety i think you want also treat the slice length computation as the critical section. or to avoid that lock, and since you aren't looking at the indices and are just checking the length, you could count how many times you read from the channel and change this if statement to `len(kzCommitments)` == `nChannelReads`. that would also clean up the writers to the channel, which could be leaking as currently written, because you could delete the `blobNotifierChan` without reading and unblocking them.